signing_service/src/tests/tests.py

```diff
@@ -0,0 +1,160 @@
+from unittest import TestCase
+import socket
+import sys
+
+import mock
+
+from ..constants import SIGNING_SERVICE_DEFAULT_PORT
+from ..constants import SIGNING_SERVICE_RECOVERABLE_ERRORS
+from ..constants import SIGNING_SERVICE_MAXIMUM_RECONNECT_TIME
+from ..signing_service import _parse_arguments
+from ..signing_service import SigningService
+
+
+class SigningServiceMainTestCase(TestCase):
+
+    def setUp(self):
+        self.host = '127.0.0.1'
+        self.port = 8000
+        self.initial_reconnect_delay = 2
+
+    def test_that_signing_service_should_be_instantiated_correctly_with_all_parameters(self):
+        signing_service = SigningService(self.host, self.port, self.initial_reconnect_delay)
+
+        self.assertIsInstance(signing_service, SigningService)
+        self.assertEqual(signing_service.host, self.host)
+        self.assertEqual(signing_service.port, self.port)
+        self.assertEqual(signing_service.initial_reconnect_delay, self.initial_reconnect_delay)
+
+    def test_that_signing_service_should_run_full_loop_when_be_instantiated_with_all_parameters(self):
+        with mock.patch('socket.socket.connect') as mock_socket_connect:
+            with mock.patch('src.signing_service.SigningService._handle_connection') as mock__handle_connection:
+                with mock.patch('socket.socket.close') as mock_socket_close:
+                    with mock.patch('src.signing_service.SigningService._was_sigterm_caught', side_effect=[False, True]):
+                        signing_service = SigningService(self.host, self.port, self.initial_reconnect_delay)
+                        signing_service.run()
+
+        mock_socket_connect.assert_called_once_with(('127.0.0.1', 8000))
+        mock_socket_close.assert_called_once()
+        mock__handle_connection.assert_called_once()
+
+    def test_that_signing_service_should_run_send_check_if_connection_is_still_alive(self):
+        with mock.patch('socket.socket.connect') as mock_socket_connect:
+            with mock.patch('src.signing_service.SigningService._handle_connection') as mock__handle_connection:
+                with mock.patch('socket.socket.send') as mock_socket_send:
+                    with mock.patch('socket.socket.close') as mock_socket_close:
+                        with mock.patch('src.signing_service.SigningService._was_sigterm_caught', side_effect=[False, False, True]):
+                            signing_service = SigningService(self.host, self.port, self.initial_reconnect_delay)
+                            signing_service.run()
+
+        mock_socket_connect.assert_called_once_with(('127.0.0.1', 8000))
+        mock_socket_send.assert_called_once_with(b'')
+        mock__handle_connection.assert_called_once()
+        self.assertEqual(mock_socket_close.call_count, 2)
```

Comment:
why is socket closed 2 times?

Reply:
`close` is run in `finally` block of the main loop. Normally it would not be be called until inner loop in `_handle_connection` is interrupted, but the whole function is mocked here.